### PR TITLE
Preview validation

### PR DESCRIPTION
**Pull request by elena-vi**

### CODE REVIEW

**Comment by elena-vi** [comment]
[Step 2 - UI preview Letter](https://trello.com/c/jjZPk7OR/154-step-2-ui-preview-letter)

File: spec/lib/hackney/service_charge/gateway/service_charge_gateway_spec.rb [hunk 1]
```diff
@@ -77,4 +77,13 @@
       end
     end
   end
+
+  context 'when retrieving fake service charge cases' do
```
Comment (alexdesi) [inline spec/lib/hackney/service_charge/gateway/service_charge_gateway_spec.rb]: "when retrieving" is already present in the external description.
we can just describe the state in which this test will run. A.i.  'when the case has a ref and uk correspondence address'

File: lib/hackney/pdf/preview_generator.rb [hunk 2]
```diff
@@ -1,24 +1,36 @@
...
+      private
+
+      def validate_mandatory_fields(letter_params)
+        MANDATORY_LETTER_FIELDS.each do |mandatory_field|
+          next if letter_params[mandatory_field].present?
+          @errors << {
+            field: mandatory_field.to_s,
+            error: 'missing mandatory field'
+          }
+        end
```
[the remark this answers is not in the document]
Comment (elena-vi) [reply lib/hackney/pdf/preview_generator.rb]: this is all rubocop thinking he knows better 🤖 
I had a simple 
```
unless letter_params[mandatory_field].present?
            @errors << {
              field: mandatory_field.to_s,
              error: 'missing mandatory field'
            }
          end
```

File: lib/hackney/service_charge/gateway/service_charge_gateway.rb [hunk 3]
```diff
@@ -80,25 +87,36 @@ def international?(postcode)
...
+              payment_ref: ref,
+              correspondence_address_1: ' 136 Southwark St',
+              correspondence_address_2: 'Southwark',
+              correspondence_address_3: 'London',
+              correspondence_postcode: 'SE1 0SW',
+              full_names_of_current_lessees: ['Soraya Clarke']
+            }
+          end
+
+          { payment_ref: ref }
```
[the remark this answers is not in the document]
Comment (elena-vi) [reply lib/hackney/service_charge/gateway/service_charge_gateway.rb]: I don't really like case statements but okay

File: spec/lib/hackney/service_charge/gateway/service_charge_gateway_spec.rb (hunk 1 above)
Comment (elena-vi) [reply spec/lib/hackney/service_charge/gateway/service_charge_gateway_spec.rb]: I disagree, there's a test context for `'when retrieving service charge cases'` and this one is specifically tied to retrieving _fake_ data

File: lib/hackney/pdf/preview_generator.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (alexdesi) [reply lib/hackney/pdf/preview_generator.rb]: - we need to configure rubocop on our needs (to discuss)

[the remark this answers is not in the document]
Comment (alexdesi) [reply lib/hackney/pdf/preview_generator.rb]: (I am thinking to a different approach)

[the remark this answers is not in the document]
Comment (alexdesi) [reply lib/hackney/pdf/preview_generator.rb]: maybe this:
```fields_with_errors = MANDATORY_LETTER_FIELDS.select{ |field| letter_params[field].nil? }
@errors = fields_with_errors.map({field: mandatory_field.to_s, error: 'missing mandatory field'})
```
or this
```
MANDATORY_LETTER_FIELDS
  .select{ |field| letter_params[field].nil? }
  .map({field: mandatory_field.to_s, error: 'missing mandatory field'})
```

File: lib/hackney/pdf/templates/letter_1_template.erb [hunk 4]
```diff
@@ -29,29 +29,29 @@
 
   <p>
     <strong>Your reference number:</strong>
-    <%= payment_ref %>
+    <%= params.fetch(:payment_ref) %>
```
Comment (alexdesi) [inline lib/hackney/pdf/templates/letter_1_template.erb]: great to wrap the params on `params`! 👍

File: lib/hackney/pdf/preview_generator.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (mrosel) [reply lib/hackney/pdf/preview_generator.rb]: then would you suggest something like this?
```ruby
if letter_params[mandatory_field].blank?
  @errors << {
    field: mandatory_field.to_s,
    error: 'missing mandatory field'
  }
end
```

File: lib/hackney/pdf/preview_generator.rb [hunk 5]
```diff
@@ -1,24 +1,39 @@
...
+          errors: @errors
+        }
+      end
+
+      private
+
+      def validate_mandatory_fields(letter_params)
+        MANDATORY_LETTER_FIELDS.each do |mandatory_field|
+          next if letter_params[mandatory_field].present?
+          @errors << {
```
Comment (mrosel) [inline lib/hackney/pdf/preview_generator.rb]: This is really good but it's not as consistent with the rest of the error handling. 
SUGGEST using `attr_reader` for errors, so you can call `.errors.any?`

and/or, I quite like how ActiveModel handles errors. https://api.rubyonrails.org/classes/ActiveModel/Errors.html

File: lib/hackney/pdf/preview_generator.rb [hunk 6]
```diff
@@ -1,24 +1,39 @@
...
+            field: mandatory_field.to_s,
+            error: 'missing mandatory field'
```
Comment (mrosel) [inline lib/hackney/pdf/preview_generator.rb]: SUGGEST using `:name` and  `:message` to be consistent

File: lib/hackney/pdf/preview_generator.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (elena-vi) [reply lib/hackney/pdf/preview_generator.rb]: Screenshot 2019-03-12 at 11 29 04
Sorry @mrosel

[the remark this answers is not in the document]
Comment (mrosel) [reply lib/hackney/pdf/preview_generator.rb]: I prefer @alexdesi suggestion on using `#map`

**Pull request merged**